docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
-    [torch.LongTensor with no dimension]
-    and values:
-    [torch.FloatTensor with no dimension]
+    >>> torch.sparse_coo_tensor(size=(2, 3))
+    tensor(indices=tensor([], size=(2, 0)),
+           values=tensor([], size=(0,)),
+           size=(2, 3), nnz=0, layout=torch.sparse_coo)
+
+.. _sparse-hybrid-coo-docs:
+
+Hybrid COO tensors
+------------------
+
+You can also construct hybrid tensors, where only the first dimensions
+are sparse, and the rest of the dimensions are dense. For instance,
+
+    >>> i = [[2, 4]]
+    >>> v = [[1, 3], [5, 7]]
+    >>> s = torch.sparse_coo_tensor(i, v, (5, 2))
+    >>> s.to_dense()
+    tensor([[0, 0],
+            [0, 0],
+            [1, 3],
+            [0, 0],
+            [5, 7]])
+
+where ``s`` is a (1+1)-dimensional :ref:`hybrid tensor
+<sparse-hybrid-docs>`:
+
+    >>> s[0]
+    tensor([0, 0])
+    >>> s[2]
+    tensor([1, 3])
+
+Notice that slicing in dense dimensions is supported (except with
+negative step):
+
+    >>> s[2, 1:]
+    tensor([3])
+
+In general, a (M + K)-dimensional coalesced sparse COO tensor ``s``
+with the number of specified elements ``nse`` has the following
+invariants:
+
+    >>> s = s.coalesce()
+    >>> M, K = s.sparse_dim(), s.dense_dim()
+    >>> assert M + K == len(s.shape) == s.ndim
+    >>> assert s.indices().shape == (M, nse)
+    >>> assert s.values().shape == (nse,) + s.shape[M:]
+
+Since ``s.indices()`` is always a 2D tensor, the smallest sparse
+dimension M = 1.  Therefore, representation of a sparse tensor of M =
+0 is simply a dense tensor.
+
+.. _sparse-uncoalesed-coo-docs:
+
+Uncoalesced sparse COO tensors
+------------------------------
+
+Torch sparse COO tensor format permits *uncoalesced* sparse tensors,
+where there may be duplicate coordinates in the indices; in this case,
+the interpretation is that the value at that index is the sum of all
+duplicate value entries:
+
+    >>> i = [[1, 1]]
+    >>> v =  [3, 4]
+    >>> s=torch.sparse_coo_tensor(i, v, (3,))
+    >>> s
+    tensor(indices=tensor([[1, 1]]),
+           values=tensor(  [3, 4]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+    >>> s.to_dense()
+    tensor([0, 7, 0])
 
-SparseTensor has the following invariants:
-  1. sparse_dim + dense_dim = len(SparseTensor.shape)
-  2. SparseTensor._indices().shape = (sparse_dim, nnz)
-  3. SparseTensor._values().shape = (nnz, SparseTensor.shape[sparse_dim:])
+Uncoalesced tensors permit us to implement certain operators more
+efficiently.  For the most part, you shouldn't have to care whether or
+not a sparse tensor is coalesced or not, as most operations will work
+identically given a coalesced or uncoalesced sparse tensor.  However,
+there are two cases in which you may need to care.
 
-Since SparseTensor._indices() is always a 2D tensor, the smallest sparse_dim = 1.
-Therefore, representation of a SparseTensor of sparse_dim = 0 is simply a dense tensor.
+First, if you repeatedly perform an operation that can produce
```

Comment:
This seems more appropriate for a section sparse tensor performance, which could be a cool section.

Reply:
Resolving as obsolete.